clcache.py

```diff
@@ -1392,17 +1406,17 @@ def printStatistics(cache):
 
 
 def resetStatistics(cache):
-    with cache.statistics as stats:
+    with cache.statistics.lock, cache.statistics as stats:
```

Comment:
If writing the statistics happens atomically now, is it really necessary to lock?

Reply:
Yes, I think so. In general writes need to be protected to avoid other processes updating the information at the same time. It is a bit different in this specific part of the code because the resulting statistics is all zeroes but we need to lock anyway because the atomic write always writes first to outputPath + ".new" and we need to prevent two processes doing so at the same time.

We could avoid the lock here by making atomicWrite write to new temporary unique files but I don't think it is worth it at the moment.

Reply:
Thinking a bit more, even if the temporary file did not always have the same name we would need to  lock. We don't want another process to write old information right after we reset the statistics.

Reply:
It's true that the way the temporary file name is currently constructed requires locking. However, I'm not sure I appreciate with your second comment yet. Can you give an example of how two concurrent clcache processes (one of which is merely resetting the statistics) might interact which yields broken statistics in case no locking is performed when resetting the statistics? The only potential issue I could think of was:

1. clcache instance A reads the statistics in order to bump a counter (e.g. increase the number of cache hits)
2. clcache instance B resets the statistics
3. clcache instance A writes out the new statistics, with an increased counter

If this were to happen, the number of cache hits at the end would not be 1 plus whatever the value was before the stats were reset.

However, from looking at the code, all the updates to the cache stats seem to follow the theme

```python
   with cache.statistics.lock, cache.statistics as stats:
            stats.registerCacheHit()
```

I.e. reading, updating and writing the cache is always an atomic operation.

Reply:
Based on your example, instance A is going to register a cache hit and instance B is going to reset the statistics (without the lock):

1. A: acquires the lock
2. A: reads current statistics, number of hits = 123
3. B: resets statistics without acquiring the lock, number of hits on disk = 0
4. A: writes updated statistics, number of hits on disk = 123 + 1

Resulting in incorrect data. If resetStatistics() acquires the lock this situation is not possible as either A or B would have to wait for the other process to finish the write.

1. A: acquires the lock
2. A: reads current statistics, number of hits = 123
3. B: about to reset statistics, blocks trying to acquire the lock
4. A: writes updated statistics, number of hits on disk = 123 + 1
5. A: releases the lock
6. B: succeeds acquiring the lock, writes all zeros in statistics
7. B: releases the lock

or

1. B: acquires the lock
2. A: about to bump cache hits, blocks trying to acquire the lock
3. B: writes all zeros in statistics file
4. B: releases the lock
5. A: succeeds acquiring the lock
6. A: reads current statistics, number of hits = 0
7. A: writes updated number of hits = 0 + 1
8. A: releases the lock